### ISSUE TITLE
Episode session reloads the wrong viewOffset

### ISSUE THREAD

**Issue by sprt** [pkkid/python-plexapi#638]
**Describe the issue**
TLDR: `Episode.reload()` always reloads the `viewOffset` from the `/library/metadata/<id>` endpoint, even when the episode was obtained through the `/status/sessions` endpoint. Both endpoints contain the `viewOffset` attribute, but with different semantics.

When obtained via a call to `PlexServer.sessions()` (`/status/sessions` endpoint), an `Episode` object contains an attribute `viewOffset` that refers to the current offset of that specific session (and which seems to be updated every 10 seconds).

However, when calling `Episode.reload()`, the object is reloaded with data from the `/library/metadata/<id>` endpoint. This endpoint also returns a `viewOffset`, but its value is almost always different from the **session** `viewOffset`, as the library `viewOffset` seems to be updated by Plex only after the user closes a player (and hence a session).

Encountered this trying to call `Episode.hasIntroMarker()` (which reloads the object) on an episode session. 

Note: I haven't tested it, but the all the above should also apply to the `Movie` class.

**Code snipppets**

```python
# 1. Action: Open an episode in Plex and seek to 60 secs in. Then close the player.
# 2. Action: Confirm that the /library/metadata/<id> endpoint reports viewOffset == 60000.
# 3. Action: Open the same episode (restarting from the beginning) in Plex again and pause it right away. Keep the player open.

# 4. Acquire an Episode object through the PlexServer.sessions() method.
server = PlexServer(...)
episode = server.sessions()[0]  # Assuming it's the first session.

# 5. Confirm that episode.viewOffset == 0.
assert episode.viewOffset == 0

# 6. Reload the episode.
episode.reload()

# 7. BUG: The library viewOffset is now reported, instead of the session one.
assert episode.viewOffset == 0  # This raises because now episode.viewOffset == 60000.
```

**Expected behavior**
Step 7 in the above section should not raise.

**Environment (please complete the following information):**
 - OS: Ubuntu 16.04
 - Plex version 1.21.1.3830
 - Python Version 3.9.0
 - PlexAPI version 4.2.0

**Comment by sprt** [comment]
To fix this bug without breaking existing clients, I suggest deprecating the `viewOffset` attribute and introducing `LibraryViewOffset` and `SessionViewOffset`, which would be updated depending on the object's `_initpath`. :smiley:

**Comment by blacktwin** [comment]
> reload() always reloads the viewOffset from the /library/metadata/<id>

Correct. When you are reloading the item, thats what you're doing, reloading the item not the item from the session. If you wanted the offset from the session you'd have to call the `sessions()` again. What is it that you're trying to do?

**Comment by sprt** [comment]
The issue isn't so much with the `reload()` method as I didn't even call it, so I don't care what it does.

The issue is that, as a black box, the `Episode` object exposes different semantics depending on whether I issued a conceptually read-only command (`hasIntroMarker()`) to it. This took me a while to debug, since as an end user, I don't concern myself with the inner workings of the library. I guess it could've been acceptable if the `viewOffset` was reloaded from the session, but the real problem here is that the **meaning** of `viewOffset` changed after I called `hasIntroMarker()`.

**Comment by blacktwin** [comment]
`hasIntroMarker` requires that the object is Full, meaning all the metadata is available. Because the objects from `sessions()` are partial objects they need to be reloaded in order to check the `hasIntroMarker` status.

> the real problem here is that the meaning of viewOffset changed after I called hasIntroMarker().

You're moving from the item (partial) in the session to the item (full) as it sits on the server. 

What is it that you're trying to do?

**Comment by Hellowlol** [comment]
Plexapi tries to be more cleaver then most people think under the hood. Some endpoint (like /session)  don't really return that much info in the raw xml response and at the same time returns some info that isn't available in any other endpoints. Some examples for this is https://github.com/pkkid/python-plexapi/blob/master/plexapi/base.py#L687-L696

So whenever you try to access a attribute that don't exist/can't figure we end up with reloading the object https://github.com/pkkid/python-plexapi/blob/master/plexapi/base.py#L377-L392 What this mean is that we will do at http requests the endpoint that has **stored** most data.

Since the session is ongoing the "correct" viewoffset isn't stored in the xml response yet. 
You can manually force a update of the viewoffset with updateProgress or updateTimeline (can't remember)

TLDR: We sometime has to HTTP requests do get the data used to create the attribute your trying to access. 

I'm pretty sure this has caught other user by surprise before. I _think_ this only happens on viewoffset. For future debugging you can turn on logging and see why a object reloads.

**Comment by sprt** [comment]
Thanks for your replies! :smiley: Indeed this confirms what I thought I found during debugging so I'm glad my understanding was correct.

> TLDR: We sometime has to HTTP requests do get the data used to create the attribute your trying to access.
>
> I'm pretty sure this has caught other user by surprise before. I think this only happens on viewoffset. For future debugging you can turn on logging and see why a object reloads.

I fully appreciate the need for reloading, it's clear there's simply no way around it sometimes given how the Plex API is structured. However I don't think a library should change an object's contract after a conceptually pure method call. It's just pretty confusing to an end user; and if others have been caught by this I think it makes sense to tweak the interface? I feel it's worth introducing the two attributes I mentioned instead of trying to achieve a 1:1 mapping with Plex's API, especially if it makes users' lives easier.

@blacktwin For my specific use case, I can work around this issue by simply saving the `viewOffset` separately for later use, so now that I understand what happens under the hood, I'm not blocked anymore. :+1:
